eval.go

```diff
@@ -75,23 +75,28 @@ func parseRemoteObject(v *runtime.RemoteObject, res interface{}) (err error) {
 		return
 	}
 
-	if v.Type == "undefined" || v.Value == nil {
+	value := v.Value
+	if value == nil {
+		// if js expression is undefined, it cannot resolve to any go type, so return ErrJSUndefined directly
+		if v.Type == "undefined" {
+			return ErrJSUndefined
+		}
 		rv := reflect.ValueOf(res)
-		// `res` should be a pointer. When the value that `res` points to is
-		// not a pointer, it can not be nil. In this case,
-		// return [ErrJSUndefined] or [ErrJSNull] respectively.
-		if rv.Kind() == reflect.Pointer && rv.Elem().Kind() != reflect.Pointer {
-			if v.Type == "undefined" {
-				return ErrJSUndefined
+		if rv.Kind() == reflect.Pointer {
+			switch rv.Elem().Kind() {
+			case reflect.Interface, reflect.Map, reflect.Pointer, reflect.Slice:
```

Comment:
```go
var f func()
err := Run(ctx, Evaluate("null", &f))
```

It seems that returning ErrJSNull in this case does not help. I think we can leave it to the `json` package.

Reply:
var f func()
err := json.Unmarshal([]byte("null"), &f) // err is nil
chan, func cannot be unserialized by json, so we should return ErrJSNull

Reply:
See the code below:

```go
var f func()
err := json.Unmarshal([]byte("null"), &f)
fmt.Println(err)
// <nil>
err = Run(ctx, Evaluate("null", &f))
fmt.Println(err)
// encountered a null value

err = json.Unmarshal([]byte("{}"), &f)
fmt.Println(err)
// json: cannot unmarshal object into Go value of type func()

err = Run(ctx, Evaluate("({})", &f))
fmt.Println(err)
// json: cannot unmarshal object into Go value of type func()
```

`null` is allowed by the json package. I don't understand why we should do something different. What is the advantage of returning `ErrJSNull` in this case?

Reply:
for array and struct type, should we also not return errors?
```go
var arr [3]int
err := json.Unmarshal([]byte(`null`), &arr) // err is nil

var s struct{}
err := json.Unmarshal([]byte(`null`), &s) // err is nil
```

Reply:
The zero value of array and struct are not nil, so we will treat them like int.